[dbnode] Enforce single Finalize on pooled objects

linasm opened the pull request:
**What this PR does / why we need it**:
Makes object pool panic on double `Finalize` of `RefCount`, tags `decoder` (which is not refcounted itself), or any other type that embeds `checked.FinalizeableOnce` struct.


**Special notes for your reviewer**:
A broader alternative to https://github.com/m3db/m3/pull/2603.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
NONE

**Does this PR require updating code package or user-facing documentation?**:
NONE

robskillington commented on src/x/checked/finalizeable_once.go:
```diff
@@ -0,0 +1,18 @@
+
+// SetFinalized sets the flag of object finalize.
+func (c *FinalizeableOnce) SetFinalized(f bool) {
+	c.finalized.Store(f)
```
I would use the underlying "Swap(value)" here and return what the old value was.

That way we can atomically see if we transitioned from false to true.

i.e. 
```go
func (c *FinalizeableOnce) SetFinalized() error {
  if oldValue := c.finalized.Swap(true); oldValue == true {
    return errAlreadyFinalized
  }
  return nil
}

func (c *FinalizeableOnce) SetUnfinalized() error {
  if oldValue := c.finalized.Swap(false); oldValue == false {
    return errNotFinalized
  }
  return nil
}
```

linasm replied:
Nice idea, changing.

robskillington commented on src/x/pool/object.go:
```diff
@@ -138,6 +148,15 @@ func (p *objectPool) Put(obj interface{}) {
+
+	if obj, ok := obj.(FinalizeableOnce); ok {
+		if obj.Finalized() {
+			p.onPoolAccessErrorFn(fmt.Errorf("double finalize, %v", obj))
```
nit: `double finalize, %v` -> `double finalize: %v` as per other errors using colon between message and value.

robskillington replied:
I would use `reflect.TypeOf(obj).String()` here. "%v" on `obj` will be fairly non-descript/not very useful.

linasm replied:
👍

robskillington commented on src/x/checked/finalizeable_once.go:
```diff
@@ -0,0 +1,18 @@
+import "go.uber.org/atomic"
+
+// FinalizeableOnce implements the logics needed to enforce single Finalize on pool.ObjectPool.
+type FinalizeableOnce struct {
```
We should probably embed this to `RefCount` struct and get this checking for all things that embed `RefCount`.

linasm replied:
Already a part of this PR.

robskillington commented on src/x/checked/finalizeable_once.go:
```diff
@@ -0,0 +1,18 @@
+package checked
```
Need license.

robskillington commented on src/x/checked/finalizeable_once_test.go:
```diff
@@ -0,0 +1,18 @@
+package checked
```
Need license.

robskillington commented on src/x/serialize/decoder.go:
```diff
@@ -35,6 +35,7 @@ var (
 )
 
 type decoder struct {
+	checked.FinalizeableOnce
```
Perhaps embed in the serialize/encoder too?

linasm replied:
Done.

linasm commented:
@robskillington I've finished addressing your feedback on this for completeness, but I am putting this PR on hold (draft) for the time being, as we are still exploring alternative approaches.
